criu/pages-compress.c

```diff
@@ -0,0 +1,208 @@
+			if (LZ4F_isError(lz4err)) {
+				pr_err("LZ4 Decompress error: %s\n", LZ4F_getErrorName(lz4err));
+				break;
+			}
+			{
+				const ssize_t res = write(ret, outbuf, outsize);
+				if (0 > res) {
+					pr_err("write error to output file, fd=%d, errno=%d\n", ret, errno);
+					break;
+				}
+			}
+			totalwrite += outsize;
+			offset = insize_orig - insize;
+			bytestoread = insize;
+			memmove(compbuf, compbuf + compbufsize - offset, offset);
+		}
+	}
+	close(ret);
+	pr_debug("decompression completed, read %lu, wrote %lu\n", (unsigned long)totalread, (unsigned long)totalwrite);
+	return 0;
+}
+
+static pthread_t decomp_thread = 0;
+
+static void *decompression_thread_routine(void *param) {
+	int ret;
+	int comp_fd = (intptr_t)param;
+	ret = decompress_image(comp_fd);
+	close(comp_fd);
+	if (0 > ret) {
+		pr_err("Failed to decompress image, ret=%d\n", ret);
+	}
+	pr_info("Decompression thread completed\n");
+	return NULL;
+}
+
+int decompression_thread_start(void) {
+	int ret;
+	int comp_fd;
+	char image_path[PATH_MAX];
+	const int dfd = get_service_fd(IMG_FD_OFF);
+
+	if (decomp_thread) {
+		pr_err("Decompression thread already started\n");
+		return -1;
+	}
+
+	pr_debug("Opening compressed image...\n");
+	sprintf(image_path, imgset_template[CR_FD_PAGES_COMP].fmt, 1);
+	comp_fd = openat(dfd, image_path, O_RDONLY, CR_FD_PERM);
+	if (0 > comp_fd) {
+		pr_warn("Can't open '%s' for dfd=%d, errno=%d\n", image_path, dfd, errno);
+		return -1;
+	}
+
+	pr_debug("Starting decompression thread...\n");
+	{
+		// Preventing processing SIGCHLD by the thread,
+		// because SIGCHLD mask is changed by CRIU's main thread during restore.
+		sigset_t blockmask, oldmask;
+		sigemptyset(&blockmask);
+		sigaddset(&blockmask, SIGCHLD);
+		if (sigprocmask(SIG_BLOCK, &blockmask, &oldmask) == -1) {
+			pr_perror("Cannot set mask of blocked signals");
+			close(comp_fd);
+			return -1;
+		}
+		ret = pthread_create(&decomp_thread, NULL, decompression_thread_routine, (void *)(intptr_t)comp_fd);
+		if (sigprocmask(SIG_SETMASK, &oldmask, NULL) == -1) {
+			pr_perror("Can not unset mask of blocked signals");
+		}
+		if (ret) {
+			pr_err("Can't start decompression thread, pthread_create returned %d\n", ret);
+			close(comp_fd);
+			return ret;
+		}
+	}
+	pr_debug("Decompression thread started\n");
+	return 0;
+}
+
+int decompression_thread_join(void) {
+	int ret;
+	if (!decomp_thread) {
+		pr_err("Decompression thread doesn't exist\n");
+		return -1;
+	}
+	ret = pthread_join(decomp_thread, NULL);
+	if (ret) {
+		pr_err("Decompression thread joining failed, ret=%d\n", ret);
+		return -1;
+	}
+	decomp_thread = 0;
+	pr_debug("Decompression thread joined\n");
+	return 0;
+}
+
+int decompression_get_fd(void) {
+	int ret;
+	pr_debug("wait for semaphore...\n");
+	ret = open(temp_path_decompressed, O_RDONLY, 0600);
+	if (0 > ret) {
+		pr_err("Failed open %s errno=%d\n", temp_path_decompressed, errno);
+	}
+	return ret;
```

Comment:
Here should be `unlink(temp_path_decompressed);` as otherwise each restore will create a leftover file `/tmp/decompressed`.
But then it would be better to rather open the file with `O_TMPFILE` and there are also no longer issues with its filename (and `mkstemp()`). Just one must keep the file descriptor instead of its close+reopen.

Reply:
I tried to minimize CRIU changes in this PR, so CRIU's behavior is still unchanged in general. Currently CRUI implementation still wants to open and close 'page-1.img' file, and it does this twice. In this PR I just follow this behaviour replacing the target file only, i.e. on CRIU demand to open 'page-1.img', it opens a decompressed file instead and returns a descriptor. 

Moreover, it was discovered that all the descriptors should be closed on 'clone' call time to avoid further problems on restore. So currently we cannot keep the decompressed file descriptor opened, because it needs to make appropriate changes in CRIU behaviour making this PR too complicated.


BTW speaking about performance when opening/closing the same file, it seems supported by system file cache good enough currently, so it doesn't seem a problem. The longest operations are decompression and further data flushing for now.

Reply:
Then one should at least handle the leftover files in /tmp and their non-unique filename leading to conflicts when multiple processes/users run CRIU.